# Домашнее задание №2

**

## Часть 1. ML workflow (**всего 5 баллов**)

In [191]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV, cross_validate

from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor

from sklearn.metrics import mean_squared_error, r2_score

from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline

from itertools import product

### Загрузим данные для работы. 

In [215]:
df = pd.read_csv("winequality-red.csv")
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

Будем решать задачу регрессии: необходимо предсказать качество вина на основе его характеристик

### Шаг 1.  (**0.2 балла**)
Создайте матрицу X объект-признак и целевой вектор y ("quality")

In [193]:
#your code here
X = df.iloc[:,:-1]
y = df['quality']

### Шаг 2. (**0.2 балла**)
Разбейте данные на train и test (доля тестовых данных - 30%).

In [194]:
#your code here
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.3, \
                                               random_state=1) 

### Шаг 3. (**0.2 балла**)
Обучите линейную регрессию на тренировочных данных и сделайте предсказания на train и на test.

In [195]:
#your code here
model = LinearRegression()
model.fit(Xtrain, ytrain)

pred_train = model.predict(Xtrain)
pred_test = model.predict(Xtest)

### Шаг 4. (**0.4 балла**)
Выведите на экран ошибку MSE на train и на test, затем выведите на экран ошибку r2 на train и test.  

In [196]:
#your code here
print("MSE_train:", mean_squared_error(ytrain,pred_train), "MSE_test:", mean_squared_error(ytest,pred_test))

MSE_train: 0.4192375220450585 MSE_test: 0.4147279618151055


In [197]:
 print("r2_test:", r2_score(ytest,pred_test),"r2_train:", r2_score(ytrain,pred_train))

r2_test: 0.3186005776026677 r2_train: 0.3732759065136656


### Шаг 5. (**0.5 балла**)
Вычислите среднее качество (r2) модели на кросс-валидации с k=5 фолдами.

In [198]:
cross_val_score(model, X, y, cv=5, scoring='r2').mean()

0.29004162884219614

### Шаг 6.  (**0.5 балла**)
Теперь примените линейную регрессию с L1-регуляризацией (Lasso) для данной задачи. Объявите модель и подберите параметр регуляризации alpha по сетке. Ищите alpha в диапазоне (0.1, 1.1) с шагом 0.1. 

Осуществите подбор параметра alpha по тренировочным данным (Xtrain, ytrain).

In [226]:
from sklearn.linear_model import Lasso
for a in np.arange(0.1, 1.1, 0.1):
    lasso = Lasso(alpha=a)
    lasso.fit(Xtrain, ytrain)
    pred_tr = lasso.predict(Xtrain)
    pred_ts = lasso.predict(Xtest)
    print('alpha={}'.format(a))
    print('Train MSE:', mean_squared_error(ytrain, pred_tr))
    print('Test MSE:', mean_squared_error(ytest, pred_ts), '\n')

alpha=0.1
Train MSE: 0.5017090419218307
Test MSE: 0.47738492434060864 

alpha=0.2
Train MSE: 0.5399545522023987
Test MSE: 0.5081972472244055 

alpha=0.30000000000000004
Train MSE: 0.584739043401611
Test MSE: 0.546596893336057 

alpha=0.4
Train MSE: 0.6388871562331493
Test MSE: 0.5939599380319246 

alpha=0.5
Train MSE: 0.6412944028004144
Test MSE: 0.5941936322663466 

alpha=0.6
Train MSE: 0.644236812174557
Test MSE: 0.5950737209467652 

alpha=0.7000000000000001
Train MSE: 0.6447200264637941
Test MSE: 0.5951623345344943 

alpha=0.8
Train MSE: 0.6448627184646102
Test MSE: 0.5951209016474258 

alpha=0.9
Train MSE: 0.6450244360655353
Test MSE: 0.5951003579191751 

alpha=1.0
Train MSE: 0.645205179266569
Test MSE: 0.5951007033497423 



### Шаг 7.  (**0.5 балла**)
Выведите наилучший алгоритм и наилучшее качество по результатам подбора alpha (best_estimator_ и best_score_).

In [200]:
#your code here
params = {'fit_intercept':[True,False], 'normalize':[True,False], 'copy_X':[True, False]}
grid = GridSearchCV(model, params)

grid.fit(Xtrain, ytrain)

print('best score:', grid.best_score_)
print('best estimator:', grid.best_estimator_)

best score: 0.3572051286705922
best estimator: LinearRegression()


### Шаг 8.  (**0.5 балла**)

С помощью найденного best_estimator_ сделайте предсказание на тестовых данных и выведите на экран r2-score на тесте.

In [201]:
#your code here
model2 = LinearRegression()
model2.fit(Xtest, ytest)
print(model2.score(Xtest, ytest))

0.3383605758141326


### Шаг 9.  (**0.5 балла**)

Попробуем улучшить качество модели за счет добавления полиномиальных признаков. Создайте pipeline, состоящий из добавления полиномиальных признаков степени 2, а затем применения линейной регрессии.

Затем вычислите r2-score этой модели на кросс валидации с пятью фолдами.

In [221]:
from sklearn.preprocessing import StandardScaler

#your code here
pipe = Pipeline([('scaler', StandardScaler()), ('polinomial', PolynomialFeatures(degree = 2)), ('mode', LinearRegression())])
cross_val_score(pipe, X, y, cv=5, scoring ='r2').mean()
gg = cross_validate(model2, X, y, cv=5, scoring='r2')
print(gg['test_score'].mean())

0.29004162884219614


### Шаг 10.  (**0.5 балла**)
Обучите модель (pipeline) на тренировочных данных и сделайте предсказания для train и test, затем выведите на экран r2-score и MSE на тренировочных и на тестовых данных.

In [237]:
#your code here
pipe.fit(Xtrain, ytrain)

pred_test1 = pipe.predict(Xtest)
pred_train1 = pipe.predict(Xtrain)

print ("R2 test:", r2_score(ytest,pred_test1), '\n', "R2 train:", r2_score(ytrain, pred_train1))
print("MSE test:", mean_squared_error( ytest , pred_test1), '\n', "MSE train:", mean_squared_error(ytrain, pred_train1))

R2 test: 0.2920650928970544 
 R2 train: 0.4591852895831121
MSE test: 0.4308785588452828 
 MSE train: 0.3617697507358328


### Сделайте выводы. Для этого ответьте на вопросы: (**1 балл**)

1) Хорошее ли качество показала исходная модель (линейная регрессия без регуляризации)? Является ли эта модель переобученной?

2) Помогла ли L1-регуляризация улучшить качество модели?

3) Помогло ли добавление полиномов второй степени улучшить качество модели? Как добавление новых признаков повлияло на переобучение?

1) Ну как минимум $R^2$ принимает положительные значения, значит какая-то взаимосвязь есть, но назвать эту модель хорошей нельзя так как $R^2$ < 0.5. Модель не переобучена, так как отличие и $R^2$ и MSE на train и test незначительное

2) MSE стал выше $\rightarrow$ модель лучше не стала

3) На тесте $R^2$ снизился $\rightarrow$ качество хуже. А разница качества между train и test выросла $\rightarrow$ переобучение. 

### *Попытайтесь улучшить модель (добейтесь наилучшего качества) - можно использовать любые методы.

При улучшении качества r2 на 0.1-0.2 +1 балл, при большем улучшении +2 балла (дополнительно к 5 баллам за основную часть).

In [ ]:
#your code here

## Часть 2. Target encoding (**всего 5 баллов**)

В этом части домашнего задания вы будете работать с выборкой `1C`. Вам нужно посчитать счетчики для `item_id` четырьмя способами:

    1) При помощи KFold схемы;  
    2) При помощи Leave-one-out схемы;
    3) При помощи smoothing схемы;
    4) При помощи expanding mean схемы.

### Подготовка данных

In [216]:
sales = pd.read_csv('sales_train.csv.gz')
sales.columns = ['date', 'date_block_num', 'shop_id', 'item_id', 'item_price', 'target']
sales

date  date_block_num  shop_id  item_id  item_price  target
0        02.01.2013               0       59    22154      999.00     1.0
1        03.01.2013               0       25     2552      899.00     1.0
2        05.01.2013               0       25     2552      899.00    -1.0
3        06.01.2013               0       25     2554     1709.05     1.0
4        15.01.2013               0       25     2555     1099.00     1.0
...             ...             ...      ...      ...         ...     ...
2935844  10.10.2015              33       25     7409      299.00     1.0
2935845  09.10.2015              33       25     7460      299.00     1.0
2935846  14.10.2015              33       25     7459      349.00     1.0
2935847  22.10.2015              33       25     7440      299.00     1.0
2935848  03.10.2015              33       25     7460      299.00     1.0

[2935849 rows x 6 columns]

In [217]:
index_cols = ['shop_id', 'item_id', 'date_block_num']

# For every month we create a grid from all shops/items combinations from that month
grid = [] 
for block_num in sales['date_block_num'].unique():
    cur_shops = sales[sales['date_block_num']==block_num]['shop_id'].unique()
    cur_items = sales[sales['date_block_num']==block_num]['item_id'].unique()
    grid.append(np.array(list(product(*[cur_shops, cur_items, [block_num]])),dtype='int32'))

#turn the grid into pandas dataframe
grid = pd.DataFrame(np.vstack(grid), columns = index_cols,dtype=np.int32)

#get aggregated values for (shop_id, item_id, month)
gb = sales.groupby(index_cols,as_index=False).agg({'target':'sum'})

#join aggregated data to the grid
all_data = pd.merge(grid,gb,how='left',on=index_cols).fillna(0)
#sort the data
all_data.sort_values(['date_block_num','shop_id','item_id'],inplace=True)

### Mean encodings без регуляризации

После проделанной технической работы, мы готовы посчитать счетчики для переменной `item_id`. 

Ниже приведены две реализации подсчета счетчиков без регуляризации. Можно использовать данный код в качестве стартовой точки для реализации различных техник регуляризации.

#### Способ 1

In [218]:
# Calculate a mapping: {item_id: target_mean}
item_id_target_mean = all_data.groupby('item_id').target.mean()

# In our non-regularized case we just *map* the computed means to the `item_id`'s
all_data['item_target_enc'] = all_data['item_id'].map(item_id_target_mean)

# Fill NaNs
all_data['item_target_enc'].fillna(0.3343, inplace=True) 

# Print correlation
encoded_feature = all_data['item_target_enc'].values
print(np.corrcoef(all_data['target'].values, encoded_feature)[0][1])

0.4830386988621791


#### Способ 2

In [219]:
'''
     Differently to `.target.mean()` function `transform` 
   will return a dataframe with an index like in `all_data`.
   Basically this single line of code is equivalent to the first two lines from of Method 1.
'''
all_data['item_target_enc'] = all_data.groupby('item_id')['target'].transform('mean')

# Fill NaNs
all_data['item_target_enc'].fillna(0.3343, inplace=True) 

# Print correlation
encoded_feature = all_data['item_target_enc'].values
print(np.corrcoef(all_data['target'].values, encoded_feature)[0][1])

0.4830386988621791


###  KFold схема (**1.25 балла**)

Необходимо реализовать Kfold схему с пятью фолдами. Используйте KFold(5) из sklearn.model_selection. 

1. Разбейте данные на 5 фолдов при помощи `sklearn.model_selection.KFold` с параметром `shuffle=False`.
2. Проитерируйтесь по фолдам: используйте 4 обучающих фолда для подсчета средних значений таргета по `item_id` и заполните этими значениями валидационный фолд на каждой итерации.

Обратите внимание на **Способ 1** из примера. В частности, изучите, как работают функции map и pd.Series.map. Они довольно полезны во многих ситуациях. 

In [240]:
# YOUR CODE GOES HERE
from sklearn.model_selection import KFold
kfold = KFold(n_splits=5, shuffle=False)
kfold1 = kfold.split(all_data)

for train, test in kfold1:
    item_id_target_mean = all_data.loc[train].groupby('item_id').target.mean()
    all_data['item_target_enc'].loc[test] = all_data['item_id'].map(item_id_target_mean)

all_data['item_target_enc'].fillna(0.3343, inplace=True)    

encoded_feature = all_data['item_target_enc'].values
# You will need to compute correlation like that
corr = np.corrcoef(all_data['target'].values, encoded_feature)[0][1]
print(corr)

0.4166330182247385


Ожидаемый ответ 0.4165

### Leave-one-out схема (**1.25 балла**)

Необходимо реализовать leave-one-out схему . Учтите, если вы запустите код из первого задания, задав количество фолдов такое же как размер выборки, то вы, вероятно, получите правильный ответ, но ждать будете очень-очень долго.

Для более быстрой реализации подсчета среднего таргета на всех объектах, кроме одного, вы можете:

1. Вычислить суммарный таргет по всем объектам.
2. Вычесть таргет конкретного объекта и разделить результат на `n_objects - 1`. 

Заметим, что пункт `1.` следует сделать для всех объектов. Также заметим, что пункт `2.` может быть реализован без циклов `for`.

Здесь может оказаться полезной функция .transform из **Способа 2** из примера.

In [246]:
# YOUR CODE GOES HERE
item_id_target_sum = all_data.groupby('item_id').target.sum()
all_data['target_sum'] = all_data['item_id'].map(item_id_target_sum)
item_id_target_count = all_data.groupby('item_id').target.count()
all_data['target_count'] = all_data['item_id'].map(item_id_target_count)
all_data['item_target_enc'] = (all_data['target_sum'] - all_data['target'])/(all_data['target_count'] - 1)

corr = np.corrcoef(all_data['target'].values, encoded_feature)[0][1]
print(corr)

0.480384831129305


Ожидаемый ответ 0.4803

### Smoothing (**1.25 балла**)

Необходимо реализовать smoothing с $\alpha = 100$. Используйте формулу:

$\frac{mean(target) \cdot nrows + globalmean \cdot \alpha }{nrows + \alpha}$,

где $globalmean=0.3343$. Заметим, что `nrows` - это количество объектов, принадлежащих конктертной категории, а не количество строк в датасете.

In [244]:
# YOUR CODE GOES HERE
globalmean=0.3343
alpha = 100

item_id_target_mean = all_data.groupby('item_id').target.mean()
all_data['item_target_mean'] = all_data['item_id'].map(item_id_target_mean)
item_id_target_count = all_data.groupby('item_id').target.count()
all_data['item_target_count'] = all_data['item_id'].map(item_id_target_count)
all_data['item_target_enc'] = (all_data['item_target_mean'] * all_data['item_target_count'] + globalmean * alpha) / (all_data['item_target_count'] + alpha)
encoded_feature = all_data['item_target_enc'].values

corr = np.corrcoef(all_data['target'].values, encoded_feature)[0][1]
print(corr)

0.4818198797097282


Ожидаемый ответ 0.4818

### Expanding mean схема (**1.25 балла**)

Необходимо реализовать *expanding mean* схему. Ее суть заключается в том, чтобы пройти по отсортированному в определенном порядке датасету (датасет сортируется в самом начале задания) и для подсчета счетчика для строки $m$ использовать строки от $0$ до $m-1$. Вам будет необходимо воспользоваться pandas функциями [`cumsum`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.core.groupby.DataFrameGroupBy.cumsum.html) и [`cumcount`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.core.groupby.GroupBy.cumcount.html).

In [ ]:
# YOUR CODE GOES HERE

corr = np.corrcoef(all_data['target'].values, encoded_feature)[0][1]
print(corr)

Ожидаемый ответ 0.5025